extension/storage/storage.go

```diff
@@ -51,6 +53,48 @@ type Client interface {
 	// Delete will delete data associated with the specified key
 	Delete(context.Context, string) error
 
+	// Batch handles specified operations in batch. Get operation results are put in-place
+	Batch(context.Context, ...Operation) error
+
 	// Close will release any resources held by the client
 	Close(context.Context) error
 }
+
+const (
+	Get = iota
+	Set
+	Delete
+)
+
+type operation struct {
+	// Key specifies key which is going to be get/set/deleted
+	Key string
```

Comment:
Food for thought:

Making this a string will enforce copying the string in many places whereas making this a `[]byte` would not. I'm talking about places like these:

```go
...
bucket.Get([]byte(op.Key))
...
```

This won't be a huge difference but might save some allocations (it depends what's the priority: interface ease of use or performance).

In the end, bbolt's implementation accepts `[]byte` as key.

Doing some ad hoc benchmark on my machine, baseline:

```
goos: darwin
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/extension/storage/filestorage
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkClientGet
BenchmarkClientGet-16             254341              9822 ns/op            5433 B/op         29 allocs/op
BenchmarkClientGet100
BenchmarkClientGet100-16          136087             15871 ns/op            7816 B/op        128 allocs/op
BenchmarkClientSet
BenchmarkClientSet-16             203229             11949 ns/op            6276 B/op         42 allocs/op
BenchmarkClientSet100
BenchmarkClientSet100-16           47556             47913 ns/op           18340 B/op        343 allocs/op
BenchmarkClientDelete
BenchmarkClientDelete-16          254775              9744 ns/op            5433 B/op         29 allocs/op
PASS
ok      github.com/open-telemetry/opentelemetry-collector-contrib/extension/storage/filestorage 13.301s
```

version with `... key []byte, ...`

```
goos: darwin
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/extension/storage/filestorage
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkClientGet
BenchmarkClientGet-16             237909              9203 ns/op            5434 B/op         29 allocs/op
BenchmarkClientGet100
BenchmarkClientGet100-16          158772             15487 ns/op            7816 B/op        128 allocs/op
BenchmarkClientSet
BenchmarkClientSet-16             194180             12017 ns/op            6268 B/op         41 allocs/op
BenchmarkClientSet100
BenchmarkClientSet100-16           49833             45893 ns/op           16739 B/op        243 allocs/op
BenchmarkClientDelete
BenchmarkClientDelete-16          252370              9527 ns/op            5433 B/op         29 allocs/op
PASS
ok      github.com/open-telemetry/opentelemetry-collector-contrib/extension/storage/filestorage 17.786s
```

which gives the following results:

```
benchstat -delta-test none old.txt new.txt
name             old time/op    new time/op    delta
ClientGet-16       9.82µs ± 0%    9.20µs ± 0%   -6.30%
ClientGet100-16    15.9µs ± 0%    15.5µs ± 0%   -2.42%
ClientSet-16       11.9µs ± 0%    12.0µs ± 0%   +0.57%
ClientSet100-16    47.9µs ± 0%    45.9µs ± 0%   -4.22%
ClientDelete-16    9.74µs ± 0%    9.53µs ± 0%   -2.23%

name             old alloc/op   new alloc/op   delta
ClientGet-16       5.43kB ± 0%    5.43kB ± 0%   +0.02%
ClientGet100-16    7.82kB ± 0%    7.82kB ± 0%    0.00%
ClientSet-16       6.28kB ± 0%    6.27kB ± 0%   -0.13%
ClientSet100-16    18.3kB ± 0%    16.7kB ± 0%   -8.73%
ClientDelete-16    5.43kB ± 0%    5.43kB ± 0%    0.00%

name             old allocs/op  new allocs/op  delta
ClientGet-16         29.0 ± 0%      29.0 ± 0%    0.00%
ClientGet100-16       128 ± 0%       128 ± 0%    0.00%
ClientSet-16         42.0 ± 0%      41.0 ± 0%   -2.38%
ClientSet100-16       343 ± 0%       243 ± 0%  -29.15%
ClientDelete-16      29.0 ± 0%      29.0 ± 0%    0.00%
```

ref:  https://stackoverflow.com/a/43470344

Reply:
That's a good find. I think it's largely a question of what the API should accept. I think keeping keys as strings simplifies a lot of thing, though making them []byte would open new possibilities